miner/bidder.go

```diff
@@ -0,0 +1,264 @@
+		chain:      chain,
+		validators: make(map[common.Address]*ethclient.Client),
+		bestWorks:  make(map[int64]*environment),
+		newBidCh:   make(chan *environment, 10),
+		exitCh:     make(chan struct{}),
+	}
+
+	if !config.Enable {
+		return b
+	}
+
+	for _, v := range config.Validators {
+		cl, err := ethclient.DialOptions(context.Background(), v.URL, rpc.WithHTTPClient(client))
+		if err != nil {
+			log.Error("Bidder: failed to dial validator", "url", v.URL, "err", err)
+			continue
+		}
+
+		b.validators[v.Address] = cl
+	}
+
+	if len(b.validators) == 0 {
+		log.Warn("Bidder: No valid validators")
+	}
+
+	return b
+}
+
+func (b *Bidder) mainLoop() {
+	defer b.wg.Done()
+
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	<-timer.C // discard the initial tick
+
+	var (
+		bidNum        int64 = 0
+		bidUntil      time.Time
+		currentHeight = b.chain.CurrentBlock().Number.Int64()
+	)
+	for {
+		select {
+		case work := <-b.newBidCh:
+			if work.header.Number.Int64() > currentHeight {
+				bidNum = 0
+				bidUntil = time.Unix(int64(work.header.Time+b.chain.Config().Parlia.Period-b.config.DelayLeftOver), 0)
+				currentHeight = work.header.Number.Int64()
+
+				if time.Now().After(bidUntil) {
+					timer.Reset(0)
+				} else {
+					timer.Reset(bidUntil.Sub(time.Now()) / time.Duration(maxBid))
+				}
+			}
+			if bidNum < maxBid && b.isBestWork(work) {
+				b.bestWorks[work.header.Number.Int64()] = work
+			}
+		case <-timer.C:
+			if b.bestWorks[currentHeight] != nil {
+				b.bid(b.bestWorks[currentHeight])
+				b.bestWorks[currentHeight] = nil
+				bidNum++
+				if bidNum < maxBid && time.Now().Before(bidUntil) {
+					timer.Reset(bidUntil.Sub(time.Now()) / time.Duration(maxBid-bidNum))
+				}
+			}
+		case <-b.exitCh:
+			return
+		}
+	}
+}
+
+func (b *Bidder) SetWallet(wallet accounts.Wallet) {
+	b.wallet = wallet
+}
+
+func (b *Bidder) registered(validator common.Address) bool {
+	_, ok := b.validators[validator]
+	return ok
+}
+
+func (b *Bidder) register(validator common.Address, url string) error {
+	if _, ok := b.validators[validator]; ok {
+		return fmt.Errorf("validator %s already registered", validator.String())
+	}
+
+	cl, err := ethclient.DialOptions(context.Background(), url, rpc.WithHTTPClient(client))
+	if err != nil {
+		log.Error("Bidder: failed to dial validator", "url", url, "err", err)
+		return err
+	}
+
+	b.validators[validator] = cl
+	return nil
+}
+
+func (b *Bidder) unregister(validator common.Address) {
+	if _, ok := b.validators[validator]; ok {
+		delete(b.validators, validator)
+	}
+}
+
+// bid notifies the next in-turn validator the work
+// 1. compute the return profit for builder based on realtime traffic and validator commission
+// 2. send bid to validator
+func (b *Bidder) bid(work *environment) {
+	var (
+		cli     = b.validators[work.coinbase]
+		parent  = b.chain.CurrentBlock()
+		bidArgs *types.BidArgs
+	)
+
+	if cli == nil {
+		log.Error("Bidder: invalid validator", "validator", work.coinbase)
```

Comment:
I think no need to be 'Error' level, it is quite common that if the in-turn validator is not integrated with this builder.

Reply:
Done